File: 1423-maximum-number-of-occurrences-of-a-substring/maximum-number-of-occurrences-of-a-substring.js
var maxFreq = function(s, maxLetters, minSize, maxSize) {
    let output = 0
    let j = 0
    let lookup = {}

    function noUniqueChars(string) {
        let lookup = new Set(string)
        return lookup.size
    }
    
    while (j <= s.length - minSize) {
        let substr = s.slice(j, j + minSize)
        if (noUniqueChars(substr) <= maxLetters) {
            if (!(substr in lookup)) lookup[substr] = 0
            lookup[substr] += 1
            output = Math.max(lookup[substr], output)
        }
        j += 1
    }

    return output
};

File: 1423-maximum-number-of-occurrences-of-a-substring/maximum-number-of-occurrences-of-a-substring.py
class Solution:
    def maxFreq(self, s: str, maxLetters: int, minSize: int, maxSize: int) -> int:
        lookup = {}
        output = 0
        j = 0

        def isValid(word):
            my_set = set(word)

            return len(my_set) <= maxLetters

        while (j <= len(s) - minSize):
            substr = s[j: j + minSize]
            
            if (isValid(substr)):
                if (substr not in lookup):
                    lookup[substr] = 0
                lookup[substr] += 1

                output = max(output, lookup[substr])

            j += 1

        return output